From: Benjamin Drung <benjamin.drung@canonical.com>
Date: Thu, 3 Sep 2026 11:13:47 +0200
Subject: fix(selinux): support umount applet from busybox

The busybox `umount` applet does not support the option
`-R`/`--recursive`.

Implement the recursive unmounting in shell by reverse iterating over
`/proc/self/mountinfo` and unmounting matching paths.

Fixes: https://github.com/dracut-ng/dracut/issues/2684
Part of: https://github.com/dracut-ng/dracut/issues/2437
Origin: upstream, https://github.com/dracut-ng/dracut/commit/e767d6a646bd8b5994e7e46505b99a5fda1e04a3
---
 modules.d/77selinux/selinux-loadpolicy.sh | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/modules.d/77selinux/selinux-loadpolicy.sh b/modules.d/77selinux/selinux-loadpolicy.sh
index 4e00859..1c34360 100755
--- a/modules.d/77selinux/selinux-loadpolicy.sh
+++ b/modules.d/77selinux/selinux-loadpolicy.sh
@@ -46,7 +46,19 @@ rd_load_policy() {
             [ -e "$NEWROOT"/.autorelabel ] && LANG=C /usr/sbin/setenforce 0
             mount --rbind /dev "$NEWROOT/dev"
             LANG=C chroot "$NEWROOT" /sbin/restorecon -R /dev
-            umount -R "$NEWROOT/dev"
+
+            # busybox does not support umount --recursive
+            local target
+            sed '1!G;h;$!d' /proc/self/mountinfo | while read -r _ _ _ _ target _; do
+                case "$target" in
+                    "$NEWROOT"/dev | "$NEWROOT"/dev/*)
+                        target=$(printf '%b_' "$target")
+                        target=${target%_}
+                        umount "$target"
+                        ;;
+                esac
+            done
+
             return 0
         fi
 
